lib/units/device/plugins/screen/stream.js

```diff
@@ -458,9 +462,71 @@ export default syrup.serial()
                 })
-                wss.on('connection', function(ws) {
-                    var id = uuidv4()
-                    var pingTimer
+                wss.on('connection', async function(ws, req) {
+                    let id = uuidv4()
+                    let pingTimer
+
+                    // Extract token from WebSocket subprotocols
+                    const token = ws.protocol.substring('access_token.'.length)
+                    const user = !!token && jwtutil.decode(token, options.secret)
+
+                    if (!token || !user) {
+                        log.warn('WebSocket connection attempt without token from %s', req.socket.remoteAddress)
+                        ws.send(JSON.stringify({
+                            type: 'auth_error'
+                            , message: 'Authentication token required'
+                        }))
+                        ws.close(1008, 'Authentication token required')
+                        return
+                    }
+
+                    const tryCheckDeviceGroup = async(fail = false) => {
+                        try {
+                            await new Promise(r => setTimeout(r, 200))
+
+                            const deviceGroup = await group.get()
+                            if (deviceGroup.email !== user?.email) {
+                                const err = 'Device used by another user'
+                                log.warn('WebSocket authentication failed for device %s: $s', options.serial, err)
+                                ws.send(JSON.stringify({
+                                    type: 'auth_error'
+                                    , message: err
+                                }))
+                                ws.close(1008, 'Authentication failed')
+                                return
+                            }
+
+                            log.info('WebSocket authenticated for device %s', options.serial)
+
+                            // Send success message
+                            ws.send(JSON.stringify({
+                                type: 'auth_success'
+                                , message: 'Authentication successful'
+                            }))
+
+                            // Sending a ping message every now and then makes sure that
+                            // reverse proxies like nginx don't time out the connection [1].
+                            //
+                            // [1] http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout
+                            pingTimer = setInterval(wsPingNotifier, 10 * 60000) // options.screenPingInterval
+                        }
+                        catch (/** @type {any} */err) {
+                            if (!fail && err instanceof NoGroupError) {
+                                await new Promise(r => setTimeout(r, 1000))
+                                return tryCheckDeviceGroup(true)
```

Comment:
I don't think it makes sense to try again if the device is not in a group. Especially if there is a `sleep(200)` already. Sounds like two hacks for a single problem that should be fixed elsewhere.